design/src/rfcs/rfc0012_bytestream_callback_api.md

```diff
@@ -0,0 +1,112 @@
+RFC: A callback API for `ByteStream`
+====================================
+
+> Status: RFC
+
+Adding a callback API to `ByteStream` will enable developers using the SDK to implement things like checksum validations and 'read progress' callbacks.
+
+## The Implementation
+
+*Note that comments starting with '//' are not necessarily going to be included in the actual implementation and are intended as clarifying comments for the purposes of this RFC.*
+
+```rust
+// Each trait method defaults to doing nothing. It's up to implementors to
+// implement one or both of the trait methods
+/// Structs and enums implementing this trait can be inserted into a `ByteStream`,
+/// and will then be called in reaction to various events during a `ByteStream`'s
+/// lifecycle.
+pub trait ByteStreamReadCallback
+// I don't really like this bound but we need it because this will be inserted into an `Inner`
+where
+    Self: Debug + Clone + PartialEq + Eq,
+{
+    /// This callback is called for each chunk **successfully** read.
+    /// If an error occurs while reading a chunk, this will not be called.
+    /// This function takes `&mut self` so that implementors may modify
+    /// an implementing struct/enum's internal state.
+    // In order to stop the compiler complaining about these empty default impls,
+    // we allow unused variables.
+    fn on_read_chunk(&mut self, #[allow(unused_variables)] chunk: &Bytes) {}
+
+    /// This callback is called once all chunks have been successfully read.
+    /// It's passed a reference to the chunks in the form of an `AggregatedBytes`.
+    /// This function takes `&mut self` so that implementors may modify an
+    /// implementing struct/enum's internal state.
+    fn finally(&mut self, #[allow(unused_variables)] aggregated_bytes: &AggregatedBytes) {}
```

Comment:
Should this return something so that checksums can be applied to trailors, or is there another mechanism to make that work?

I don't think this can take a reference to all the data, especially in cases where the data downloaded is too large to store in memory.

Reply:
What is useful to pass here then? Just the total amount of bytes read? The number of chunks?

Reply:
In the newest design, `finally` can be optionally used to return errors. This would be used in order to return an error if a response body's checksum was requested but it didn't match what was expected. I'm open to a name change, though, if you have suggestions.